typing/parmatch.ml

```diff
@@ -1008,35 +1139,45 @@ let rec satisfiable pss qs = match pss with
           satisfiable pss (q::qs)
     | {pat_desc = (Tpat_any | Tpat_var(_))}::qs ->
         let pss = simplify_first_col pss in
-        let q0 = discr_pat omega pss in
-        begin match build_specialized_submatrices ~extend_row:(@) q0 pss with
-        | { default; constrs = [] } ->
-          (* first column of pss is made of variables only *)
-          satisfiable default qs
-        | { default; constrs }  ->
-            if full_match false constrs then
-              List.exists
-                (fun (p,pss) ->
-                  not (is_absent_pat p) &&
-                  satisfiable pss (simple_match_args p omega @ qs))
-                constrs
-            else
+        if not (first_column_is_coherent pss) then (
+          false
+        ) else (
+          let q0 = discr_pat omega pss in
+          match build_specialized_submatrices ~extend_row:(@) q0 pss with
+          | { default; constrs = [] } ->
+              (* first column of pss is made of variables only *)
               satisfiable default qs
-        end
+          | { default; constrs }  ->
+              if full_match false constrs then
+                List.exists
+                  (fun (p,pss) ->
+                    not (is_absent_pat p) &&
+                    satisfiable pss (simple_match_args p omega @ qs))
+                  constrs
+              else
+                satisfiable default qs
+        )
     | {pat_desc=Tpat_variant (l,_,r)}::_ when is_absent l r -> false
     | q::qs ->
         let pss = simplify_first_col pss in
         let q0 = discr_pat q pss in
-        satisfiable (build_specialized_submatrix ~extend_row:(@) q0 pss)
-          (simple_match_args q0 q @ qs)
+        if not (first_column_is_coherent pss) then (
```

Comment:
Here I think that we should check `first_column_is_coherent (q0 :: pss)` -- that's the idea anyway, I suspect that the types don't match and that the API would have to change slightly.